src/ir/expr.rs

```diff
@@ -47,6 +47,41 @@ impl ExprIdx {
         }
     }
 
+    pub fn resolve(
+        self,
+        ctx: &mut impl Ctx<Expr>,
+        binding: &HashMap<ParamIdx, u64>,
+    ) -> ExprIdx {
```

Comment:
I think @gabizon103's mono PR also adds something like this? Also, have you looked at `utils::Subst`? That structure provides an `apply()` method to do this kind of resolving for you. Let me know if you have trouble understanding how it's used---it's kind of tricky.